lib/coarse_op.cuh

```diff
@@ -624,58 +172,287 @@ namespace quda {
+          tp.grid.z = (arg.vuvTile.N_tiles + tp.block.z - 1) / tp.block.z;
+
+          arg.shared_atomic = tp.aux.y;
+          arg.parity_flip = tp.aux.z;
+
+          if (arg.shared_atomic) {
+            // check we have a valid problem size for shared atomics
+            // constraint is due to how shared memory initialization and global store are done
+            int block_size = arg.fineVolumeCB / arg.coarseVolumeCB;
+            if (block_size / 2 < Arg::coarseSpin * Arg::coarseSpin)
+              errorQuda("Block size %d not supported in shared-memory atomic coarsening", block_size);
+
+            arg.aggregates_per_block = tp.aux.x;
+            tp.block.x *= tp.aux.x;
+            tp.grid.x /= tp.aux.x;
+          }
+
+          if (arg.coarse_color_wave) {
+            // swap x and y grids
+            std::swap(tp.grid.y, tp.grid.x);
+            // augment x grid with coarseColor row grid (z grid)
+            arg.grid_z = tp.grid.z;
+            arg.coarse_color_grid_z = arg.vuvTile.M_tiles * tp.grid.z;
+            tp.grid.x *= tp.grid.z;
+            tp.grid.z = 1;
+          }
+
+          // this will ensure we pass the generic kernel bounds check
+          arg.threads.x = tp.grid.x * tp.block.x;
+          resizeVector(tp.grid.y * tp.block.y, tp.grid.z * tp.block.z);
+          launch_device<compute_vuv>(tp, stream, arg);
+          arg.threads.x = minThreads();
+          resizeVector((arg.parity_flip ? 1 : 2) * arg.max_height_tiles_per_block, arg.max_width_tiles_per_block);
+
+          if (arg.coarse_color_wave) {
+            // revert the grids
+            tp.grid.z = arg.grid_z;
+            tp.grid.x /= tp.grid.z;
+            std::swap(tp.grid.x,tp.grid.y);
+          }
+
+          if (arg.shared_atomic) {
+            tp.block.x /= tp.aux.x;
+            tp.grid.x *= tp.aux.x;
+          }
+
+        } // if use_mma
+
+      } else if (type == COMPUTE_COARSE_CLOVER) {
+
+#if defined(WILSONCOARSE)
+        launch_device<compute_coarse_clover>(tp, stream, arg);
+#else
+        errorQuda("compute_coarse_clover not enabled for non-wilson coarsenings");
+#endif
+
+      } else if (type == COMPUTE_REVERSE_Y) {
+        launch_device<reverse>(tp, stream, arg);
+      } else if (type == COMPUTE_DIAGONAL) {
+#if defined(WILSONCOARSE) || defined(COARSECOARSE)
+        launch_device<add_coarse_diagonal>(tp, stream, arg);
+#else
+        errorQuda("add_coarse_diagonal not enabled for staggered coarsenings");
+#endif
+      } else if (type == COMPUTE_STAGGEREDMASS) {
+#if defined(STAGGEREDCOARSE)
+        launch_device<add_coarse_staggered_mass>(tp, stream, arg);
+#else
+        errorQuda("add_coarse_staggered_mass not enabled for non-staggered coarsenings");
+#endif
+      } else if (type == COMPUTE_TMDIAGONAL) {
+#if defined(WILSONCOARSE)
+        launch_device<add_coarse_tm>(tp, stream, arg);
+#else
+        errorQuda("add_coarse_tm not enabled for non-wilson coarsenings");
```

Comment:
@maddyscientist I hit this when doing 3-level MG with tm(-clover) fermions. Is this expected?

Reply:
I saw that you've modified this to `#if defined(WILSONCOARSE) || defined(COARSECOARSE)`. I did the same locally shortly after reporting here and can confirm that this works correctly. Thanks!